lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -3877,15 +3874,50 @@ prop_balanceTransactionValid wallet@(Wallet' _ walletUTxO _) (ShowBuildable part
                 ]
         counterexample msg $ property (size <= limit)
 
+    prop_txCollBalance
+        :: Cardano.Tx era
+        -> Cardano.UTxO era
+        -> Property
+    prop_txCollBalance cardanoTx' cardanoUTxO = Write.withConstraints era $
+        let
+            tx = Write.fromCardanoTx cardanoTx'
+            body = Write.txBody era tx
+
+            lookupOut
+                :: Write.TxIn
+                -> Write.TxOut (Write.ShelleyLedgerEra era)
+            lookupOut i = fromMaybe err $ Write.txinLookup i utxo
+              where
+                err = error "prop_txCollBalance: unable to lookup input"
+                utxo = Write.fromCardanoUTxO cardanoUTxO
+
+            mtotalCol = view Write.totalCollateralBabbageTxBodyL body
+            colIn = foldMap (Write.txOutValue era . lookupOut)
+                $ Set.toList
+                $ view Write.collateralInputsBabbageTxBodyL body
+            mcolRet = Write.txOutValue era
+                <$> view Write.collateralReturnBabbageTxBodyL body
+        in
+            case (mtotalCol, mcolRet) of
+                (SJust totalCol, SJust colRet) ->
+                    colIn <-> colRet === (Value.inject totalCol)
```

Comment:
`Data.Group.~~` ?

Reply:
I'd probably lean towards using the same thing as the ledger though 🤔 It both leads us to other convenient helpers in the `Val` class and makes reading ledger code easier for us